pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,361 @@
+ * 5-Delays reported for partitions that do not have events to consume are reported as zero.
+ * 6-We track the time at which each delay sample was collected so that delay can be increased when partition stops
+ *   consuming for any reason other than no events being available for consumption.
+ * 7-Partitions whose Segments go from CONSUMING to DROPPED state stop being tracked so their delays do not cloud
+ *   delays of active partitions.
+ * 8-When a segment goes from CONSUMING to ONLINE, we start a timeout for the corresponding partition.
+ *   If no consumption is noticed after the timeout, we then read ideal state to confirm the server still hosts the
+ *   partition. If not, we stop tracking the respective partition.
+ * 9-A timer thread is started by this object to track timeouts of partitions and drive the reading of their ideal
+ *  state.
+ *
+ *  The following diagram illustrates the object interactions with main external APIs
+ *
+ *     (CONSUMING -> ONLINE state change)
+ *             |
+ *      markPartitionForConfirmation(partitionId)
+ *            |                         |<-updateIngestionDelay()-{LLRealtimeSegmentDataManager(Partition 0}}
+ *            |                         |
+ * ___________V_________________________V_
+ * |           (Table X)                |<-updateIngestionDelay()-{LLRealtimeSegmentDataManager(Partition 1}}
+ * | IngestionDelayTracker              |           ...
+ * |____________________________________|<-updateIngestionDelay()-{LLRealtimeSegmentDataManager (Partition n}}
+ *              ^                      ^
+ *              |                       \
+ *   timeoutInactivePartitions()    stopTrackingPartitionIngestionDelay(partitionId)
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|
+ *
+ * TODO: handle bug situations like the one where a partition is not allocated to a given server due to a bug.
+ */
+
+public class IngestionDelayTracker {
+
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 300000; // 5 minutes +/- precision in timeouts
+  // Once a partition is marked for verification, we wait 10 minutes to pull its ideal state.
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this amount of time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 100;
+  private static final Logger _logger = LoggerFactory.getLogger(IngestionDelayTracker.class.getSimpleName());
+
+  /*
+   * Class to keep an ingestion delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase ingestion delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  static private class DelayMeasure {
```

Comment:
Since we always count the diff with `(currentTime - sampleTime) + delayMs`, where `delayMs = sampleTime - ingestionTime`, why not directly tracking `ingestionTime`?

Reply:
^^

Reply:
Good simplification, I am working on this. I think this complexity came out of first considering measures with and not age in the same code base and then forgetting to simplify, nice catch. Working on this now